src/Agent.Worker/Build/FileContainerServer.cs

```diff
@@ -0,0 +1,149 @@
+﻿using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.FileContainer.Client;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
+{
+    public class FileContainerServer
+    {
+        private readonly ConcurrentQueue<string> _fileUploadQueue = new ConcurrentQueue<string>();
+        private readonly ConcurrentBag<Exception> _exceptionsDuringFileUpload = new ConcurrentBag<Exception>();
+        private CancellationTokenSource _uploadCancellationTokenSource;
+
+        private Uri _projectCollectionUrl;
+        private VssCredentials _credential;
+        private Guid _projectId;
+        private long _containerId;
+        private string _containerPath;
+
+        private int _filesUploaded = 0;
+        private string _sourceParentDirectory;
+
+        private FileContainerHttpClient FileContainerHttpClient { get; }
+
+        public FileContainerServer(
```

Comment:
It looks like you could increase testability of this class and of the consumers of this class. Consider adding a separate FileContainerManager class which contains the bulk of the logic here, and leave the FileContainerServer class as a thin wrapper over the http client.

Otherwise, doesn't the http client require disposing since you're making a new one each time this class is created? If FileContainerServer is a thin wrapper over the http client *and* a singleton, then we don't really need to worry about disposing it.